source/common/upstream/cds_api_impl.cc

```diff
@@ -38,13 +38,24 @@ CdsApiImpl::CdsApiImpl(const envoy::config::core::v3::ConfigSource& cds_config,
 
 void CdsApiImpl::onConfigUpdate(const std::vector<Config::DecodedResourceRef>& resources,
                                 const std::string& version_info) {
-  ClusterManager::ClusterInfoMap clusters_to_remove = cm_.clusters();
+  auto all_existing_clusters = cm_.clusters();
+  absl::flat_hash_set<absl::string_view> clusters_to_remove(
+      all_existing_clusters.active_clusters_.size() +
+      all_existing_clusters.warming_clusters_.size());
+
+  for (const auto& [name, _] : all_existing_clusters.active_clusters_) {
+    clusters_to_remove.emplace(name);
+  }
+  for (const auto& [name, _] : all_existing_clusters.warming_clusters_) {
+    clusters_to_remove.emplace(name);
+  }
```

Comment:
perf nit: is it possible to do this without building a new set? Can't you just check both maps for resources to erase below and then go from there?

Reply:
Hmm, active_clusters_ + warming_clusters_ gives a super set, and we need to remove the clusters  from `const std::vector<Config::DecodedResourceRef>& resources`
To avoid O(N^2) compare, a new set is needed anyway.

either the union of (active_clusters_ + warming_clusters_) or to_set(resources)

Reply:
```
  auto all_existing_clusters = cm_.clusters();  
  for (const auto& resource : resources) {
    all_existing_clusters.active_clusters_.erase(resource.get().name());
    all_existing_clusters.warming_clusters_.erase(resource.get().name());
  }
  Protobuf::RepeatedPtrField<std::string> to_remove_repeated;
  for (const auto& [cluster_name, _] : all_existing_clusters.active_clusters) {
    *to_remove_repeated.Add() = cluster_name;
  }
  for (const auto& [cluster_name, _] : all_existing_clusters.warming_clusters) {
if (all_existing_clusters.active_clusters_.count(cluster_name) == 0) {
    *to_remove_repeated.Add() = cluster_name;
}
  }
```

Reply:
Yeah that's kind of what I was thinking, but I don't feel strongly about it. What do you think is better? Also, can you merge main to make sure there is no conflict from the change I just merged?

/wait

Reply:
Fine to me. Merging and updating to above